### PR TITLE
Initial implementation of variadic lax.reduce()

### PR DESCRIPTION
**Pull request by jakevdp**
This leaves a couple pieces unimplemented, including support for broadcasted inputs and batch rule.

### CODE REVIEW

File: jax/lax/lax.py [hunk 1]
```diff
@@ -1020,25 +1020,40 @@ def transpose(operand: Array, permutation: Sequence[int]) -> Array:
   else:
     return transpose_p.bind(operand, permutation=permutation)
 
-def reduce(operand: Array, init_value: Array, computation: Callable,
-           dimensions: Sequence[int]) -> Array:
+def reduce(operand: Union[Array, Tuple[Array]], init_value: Union[Array, Tuple[Array]],
```
Comment (mattjj) [inline jax/lax/lax.py:1023]: Should we be polymorphic between list and tuple here, as in e.g. `Union[Array, Tuple[Array], List[Array]]`? I don't have a strong preference, but I've found users often expect tuple/list polymorphism.

File: jax/lax/lax.py [hunk 2]
```diff
@@ -1020,25 +1020,40 @@ def transpose(operand: Array, permutation: Sequence[int]) -> Array:
...
-  monoid_reducer = _get_monoid_reducer(computation, init_value)
-  if monoid_reducer:
-    return monoid_reducer(operand, dimensions)
+  return_tuple = isinstance(operand, tuple)
+  if not return_tuple:
+    operand = (operand,)
+  if not isinstance(init_value, tuple):
+    init_value = (init_value,)
+  if len(operand) != len(init_value):
+    raise TypeError("reduce: length of operand must match length of init_value; got "
```
Comment (mattjj) [inline jax/lax/lax.py:1035]:
```suggestion
    raise TypeError("reduce: length of operands tuple must match length of init_value tuple; got "
```

**Review by mattjj — APPROVED** [review]
Awesome work!

File: jax/lax/lax.py (hunk 1 above)
Comment (jakevdp) [reply jax/lax/lax.py:1023]: Good point... what do you think about `Union[Array, Sequence[Array]]` with an `isinstance(operand, typing.Sequence)` check? That returns False for numpy & jax arrays, and True for Tuples and Lists.

Comment (mattjj) [reply jax/lax/lax.py:1023]: SGTM! I didn't know that kind of `isinstance` check worked.

**Pull request merged**